digital_land/commands.py

```diff
@@ -531,6 +533,220 @@ def collection_add_source(entry, collection, endpoint_url, collection_dir):
+
+                    try:
+                        given_start_date = datetime.strptime(
+                            row["start-date"], "%Y-%m-%d"
+                        ).date()
+                    # need to raise ValueError here otherwise datetime will raise it's own error, not the clear format we want
+                    except ValueError:
+                        raise ValueError(
+                            f"start-date {given_start_date} must be format YYYY-MM-DD"
+                        )
+
+                    if given_start_date > datetime.today().date():
+                        raise ValueError(
+                            f"The start_date {given_start_date} cannot be in the future"
+                        )
+                except ValueError:
+                    raise
+                except Exception as e:
+                    raise ValueError(f"An unexpected error occured: {str(e)}")
+
+            if (
+                row["organisation"] == ""
+                or row["organisation"] not in valid_organisations
+            ):
+                if row["organisation"] == "":
+                    raise ValueError(
+                        f"The given organisation '{row['organisation']}' is blank"
+                    )
+                if row["organisation"] not in valid_organisations:
+                    raise ValueError(
+                        f"The given organisation '{row['organisation']}' is not in our valid organisations"
+                    )
+
+            # do we need a check to see if the given pipeline exists?
+        # SHOULD HAVE ALL CHECKS DONE, NOW CREATE HASH FOR ENDPOINT AND
+    print("======================================================================")
+    print("Endpoint and source details")
+    print("======================================================================")
+    print("Endpoint URL: ", row["endpoint-url"])
+    print("Endpoint Hash:", hash_value(row["endpoint-url"]))
+    print("Documentation URL: ", row["documentation-url"])
+    print()
+
+    # if endpoint already exists, it will indicate it and quit function here
+    if collection.add_source_endpoint(row):
+        endpoint = {
+            "endpoint-url": row["endpoint-url"],
+            "endpoint": hash_value(row["endpoint-url"]),
+            "end-date": row.get("end-date", ""),
+            "plugin": row.get("plugin"),
+            "licence": row["licence"],
+        }
+        endpoints.append(endpoint)
+    # if passes above code we can now tr the 'downlaoad resource'part
+
+    # endpoints have been added now lets collect the resources using the endpoint information
+    collector = Collector(collection_dir=collection_dir)
+
+    for endpoint in endpoints:
+        collector.fetch(
+            url=endpoint["endpoint-url"],
+            endpoint=endpoint["endpoint"],
+            end_date=endpoint["end-date"],
+            plugin=endpoint["plugin"],
+        )
```

Comment:
This method actually returns the status from the fetch, you could use it to build a map of endpoint:status. I'm wondering if we can also change it to additionally return the resource and also store this in a mapping of endpoint:resource? and then we can massively simplify the later loops when printing the log and resource info

Reply:
On second thought let me have another look at this before we decide to do it (i'm not sure the status the function is returning is exactly what we want)

Reply:
I've made a branch [here](https://github.com/digital-land/digital-land-python/commit/958a7c6b2e2ef25fa1e897b8dd296e24b4336144) showing how I think we can add the log path to the return value of the Collector .fetch function to reduce how much code we have to write